### PR TITLE
Enhancements vans

### PR DESCRIPTION
**Pull request by tonielgeitani**
# Description of the problem

-New void fraction scheme was not reflecting good values at boundaries.
-Some drag models had issues with their units
-Added a feature to print particle velocity at every time step for the sedimentation test case and output it in a file.

# Description of the solution

- Removed the need for special calculation at the boundaries for the new void fraction scheme and added the quadrature weights in the calculation of the void fraction. This way every quadrature point will have a different contribution to the cell void fraction depending on its relative weight.
- Fixed the units for the drag models and used formulations that are presented in article for the momentum exchange coefficient

# How Has This Been Tested?

- For the void fraction it has been tested by comparing different void fraction schemes for the spouted bed and finding that treating the boundaries indifferently led to the most realistic void fraction scheme and mass conservation.
- For the drag models, the spouted have been simulated with these models and the first bubble formed has been examined. Previously, the bubble was very small and incomparable to the article experiment and simulations. Now its form is completely comparable.

# Documentation

-N/A

# Future changes

-N/A

# Comments

-N/A

### CODE REVIEW

File: source/fem-dem/cfd_dem_coupling.cc [hunk 1]
```diff
@@ -1267,7 +1267,22 @@ CFDDEMSolver<dim>::print_particles_summary()
         << "--------------------------------------------------------------------------"
         << "--------------------------------------------------------------------------"
         << std::endl;
+
+      particle_velocity_table.add_value(
```
Comment (voferreira) [inline source/fem-dem/cfd_dem_coupling.cc:1271]: I would use a separate member and use a parameter to do this. I think it would be cleaner this way :)

**Review by voferreira — CHANGES_REQUESTED** [review]
Nice catch on the Drag models! Good job on cleaning the void fraction part as well. I still want to double-check the formulation and I must finish it tomorrow.

File: include/fem-dem/cfd_dem_coupling.h [hunk 2]
```diff
@@ -319,6 +319,7 @@ class CFDDEMSolver : public GLSVANSSolver<dim>
   DoFHandler<dim> background_dh;
   PVDHandler      grid_pvdhandler;
   PVDHandler      particles_pvdhandler;
+  TableHandler    particle_velocity_table;
```
Comment (blaisb) [inline include/fem-dem/cfd_dem_coupling.h:322]: ifyou are going to print all the particle velocity, will not print everything related to the particles? Shahab has already made a post-processor for this. it's called print xyz or something like this and it is already used in the DEM. No need to make a new one :)

File: source/fem-dem/cfd_dem_coupling.cc [hunk 3]
```diff
@@ -1267,7 +1267,22 @@ CFDDEMSolver<dim>::print_particles_summary()
...
     }
+
+  {
+    std::string filename =
+      this->simulation_parameters.simulation_control.output_folder +
+      "particle_velocity" + ".dat";
+    std::ofstream output(filename.c_str());
+    particle_velocity_table.set_precision("time", 12);
+    particle_velocity_table.set_precision("particle-velocity", 12);
+    particle_velocity_table.write_text(output);
```
Comment (blaisb) [inline source/fem-dem/cfd_dem_coupling.cc:1284]: Ohh I see, you wanted to write it to a file?  Then I guess that's ok, but I would add more than the y component of the velocity.
I am not sure about adding another parameter. I'm wondering what is the use of this file other than for a few comparison? Because you can get the particles velocity using the pvd and the vtu files, so what's the point of having them in a text file?

File: source/fem-dem/gls_vans.cc [hunk 4]
```diff
@@ -565,77 +565,50 @@ GLSVANSSolver<dim>::quadrature_centered_sphere_method(bool load_balance_step)
...
-                          // Particle completely in cell (PCM)
-                          particle_properties
-                            [DEM::PropertiesIndex::volumetric_contribution] +=
-                            M_PI *
-                            pow(particle_properties[DEM::PropertiesIndex::dp],
-                                dim) /
-                            (2.0 * dim);
-                        }
-                    }
+                            else if (dim == 3)
```
Comment (blaisb) [inline source/fem-dem/gls_vans.cc:603]: Wow this is much cleaner now.

File: source/fem-dem/vans_assemblers.cc [hunk 5]
```diff
@@ -1093,21 +1093,21 @@ GLSVansAssemblerKochHill<dim>::calculate_particle_fluid_interactions(
...
+                8.16 * pow(1 - cell_void_fraction, 3));
         }
       else if ((1 - cell_void_fraction) >= 0.4)
         {
           f0 = 10 * (1 - cell_void_fraction) / pow(cell_void_fraction, 3);
         }
 
       double f3 = 0.0673 + 0.212 * (1 - cell_void_fraction) +
-                  0.0232 / pow(cell_void_fraction, 5);
+                  0.0232 * pow(cell_void_fraction, -5);
```
Comment (blaisb) [inline source/fem-dem/vans_assemblers.cc:1104]: why this change?

File: source/fem-dem/vans_assemblers.cc [hunk 6]
```diff
@@ -1310,8 +1306,6 @@ GLSVansAssemblerGidaspow<dim>::calculate_particle_fluid_interactions(
 
       particle_number += 1;
     }
-
-  beta_drag = beta_drag / scratch_data.cell_volume;
```
[the remark this answers is not in the document]
Comment (blaisb) [reply source/fem-dem/vans_assemblers.cc:1314]: I'm asking myself the same question here.

**Review by blaisb — CHANGES_REQUESTED** [review]
First batch of comments, but I will need to relook at the drag expressions. Do you think you could put the reference for the formulas in the documentation so at least we all agree we are reading from the same source? :)

File: source/fem-dem/cfd_dem_coupling.cc (hunk 3 above)
Comment (voferreira) [reply source/fem-dem/cfd_dem_coupling.cc:1284]: I agree. There is no reason to have it in a separate file. If you want, I can help you to do it in the postprocessing using pyvista.

File: source/fem-dem/vans_assemblers.cc (hunk 5 above)
Comment (voferreira) [reply source/fem-dem/vans_assemblers.cc:1104]: I think I would keep the division here, just because it is easier to compare with the article.

File: source/fem-dem/vans_assemblers.cc [hunk 7]
```diff
@@ -1273,35 +1272,32 @@ GLSVansAssemblerGidaspow<dim>::calculate_particle_fluid_interactions(
           C_d = 0.44;
         }
 
-      if (cell_void_fraction >= 0.8)
+      if (cell_void_fraction > 0.8)
```
[the remark this answers is not in the document]
Comment (tonielgeitani) [reply source/fem-dem/vans_assemblers.cc:1275]: Yes this is a typo. I removed it by mistake.

File: source/fem-dem/vans_assemblers.cc (hunk 6 above)
[the remark this answers is not in the document]
Comment (tonielgeitani) [reply source/fem-dem/vans_assemblers.cc:1314]: The term of the beta drag in all the drag models should be 1/s to be equal to the units in the VANS equation. Before the momentum_transfer_coefficient unit was m^3/s so we had to divide by the element volume to get 1/s. In the Gidaspow model, the momentum_transfer_coefficient unit is already 1/s . So dividing beta by the element volume would result in the beta unit to be 1/(m^3.s) and this was leading to a very very small initial bubble in the fluidized bed. It is just a matter of units

File: source/fem-dem/vans_assemblers.cc [hunk 8]
```diff
@@ -1156,8 +1156,7 @@ GLSVansAssemblerBeetstra<dim>::calculate_particle_fluid_interactions(
   Assert(
     !scratch_data.properties_manager.is_non_newtonian(),
     RequiresConstantViscosity(
-      "GLSVansAssemblerBeetstra<dim>::calculate_particle_fluid_interactions"));
-  const double viscosity = scratch_data.properties_manager.viscosity_scale;
+      "GLSVansAssemblerDiFelice<dim>::calculate_particle_fluid_interactions"));
```
[the remark this answers is not in the document]
Comment (tonielgeitani) [reply source/fem-dem/vans_assemblers.cc:1159]: Done

File: source/fem-dem/cfd_dem_coupling.cc (hunk 3 above)
Comment (tonielgeitani) [reply source/fem-dem/cfd_dem_coupling.cc:1284]: True I only had it for the comparison of the sedimentation velocity of different void fraction models. This can be removed entirely. its useless except in this case. I will remove it then.

File: include/fem-dem/cfd_dem_coupling.h (hunk 2 above)
Comment (tonielgeitani) [reply include/fem-dem/cfd_dem_coupling.h:322]: I will remove it. It was just for one single case

File: source/fem-dem/cfd_dem_coupling.cc (hunk 1 above)
Comment (tonielgeitani) [reply source/fem-dem/cfd_dem_coupling.cc:1271]: I will remove it. It was only needed for one comparison case

File: source/fem-dem/vans_assemblers.cc (hunk 5 above)
Comment (tonielgeitani) [reply source/fem-dem/vans_assemblers.cc:1104]: Done

**Comment by tonielgeitani** [comment]
I removed the table and file creation for particle velocity as it was only needed for a single comparison case.  These are the equations for the drag:  [Beestra](https://www.sciencedirect.com/science/article/pii/S0301932220305346 ) eq. 7 and 8. They reference the main Beestra paper but they write the drag for in a way that can be written as F=Beta(u-v). All the other equations remained the same as the references from the documentation of the parameters of Unresolved CFD-DEM.

File: source/fem-dem/vans_assemblers.cc [hunk 9]
```diff
@@ -1179,26 +1178,27 @@ GLSVansAssemblerBeetstra<dim>::calculate_particle_fluid_interactions(
...
-                     -(1 + 4 * (1 - cell_void_fraction)) * 0.5));
-
-      drag_force = normalized_drag_force * 3 * M_PI * viscosity * density *
-                   particle_properties[DEM::PropertiesIndex::dp] *
-                   superficial_velocity;
-
-      beta_drag +=
-        drag_force.norm() /
-        (density * (relative_velocity[particle_number].norm()) + 1e-15);
+      // Beetstra drag coefficient
```
Comment (voferreira) [inline source/fem-dem/vans_assemblers.cc:1181]: Since you are reframing the drag model, the comment in the .h file needs to be changed as well. I would add the reference in the .h and put a note on this at the documentation.

File: source/fem-dem/vans_assemblers.cc (hunk 6 above)
[the remark this answers is not in the document]
Comment (blaisb) [reply source/fem-dem/vans_assemblers.cc:1314]: Ohh I see, so the original correlation does not lead to the same unit by definition? Interesting. It's true that the unit of the Navier-Stokes equation we solve are Velocity / time so L^1 / T^2. Since beta is multiplied by the velocity, it needs to have units of seconds inverse.

**Review by blaisb — APPROVED** [review]
Good explanations, everything makes sense to me

File: source/fem-dem/vans_assemblers.cc (hunk 9 above)
Comment (tonielgeitani) [reply source/fem-dem/vans_assemblers.cc:1181]: Done

**Comment by voferreira** [comment]
Nice catch with this PR, @tonielgeitani ! All is good to me now.

**Review by voferreira — APPROVED** [review]
